lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py

```diff
@@ -307,6 +314,114 @@ def detach(self, ue_id, reason_type, wait_for_s1_ctxt_release=True):
         with self._lock:
             del self._ue_ip_map[ue_id]
 
+    def _verify_dl_flow(self, dl_flow_rules=None):
+        # try at least 5 times before failing as gateway
+        # might take some time to install the flows in ovs
+
+        # Verify the total number of DL flows for this UE ip address
+        num_dl_flows = 1
+        for key, value in dl_flow_rules.items():
+            ipv4_src_addr = None
+            tcp_src_port = 0
+            ip_proto = 0
+            ue_ip_str = str(key)
+            # Set to 1 for the default bearer
+            total_num_dl_flows_to_be_verified = 1
+            for item in value:
+                for flow in item:
+                    if flow["direction"] == "DL":
+                        total_num_dl_flows_to_be_verified += 1
+            total_dl_ovs_flows_created = get_flows(
+                self.datapath,
+                {
+                    "table_id": self.SPGW_TABLE,
+                    "match": {
+                        "nw_dst": ue_ip_str,
+                        "eth_type": 2048,
+                        "in_port": self.LOCAL_PORT,
+                    },
+                },
+            )
+            assert (
+                len(total_dl_ovs_flows_created)
+                == total_num_dl_flows_to_be_verified
+            )
+
+            # Now verify the rules for every flow
+            for item in value:
+                for flow in item:
+                    if flow["direction"] == "DL":
+                        ipv4_src_addr = str(flow["ipv4_src"])
+                        tcp_src_port = flow["tcp_src_port"]
+                        ip_proto = (
+                            FlowMatch.IPPROTO_TCP
+                            if (flow["ip_proto"] == "TCP")
+                            else FlowMatch.IPPROTO_UDP
+                        )
```

Comment:
protocol does not need to be either TCP or UDP. It may be better to directly use enums defined in FlowMatch rather than strings in the rule definitions.

Reply:
So you are saying instead of defining "ip_proto": "TCP" in the UL/DL flow rules use "ip_proto":  IPPROTO_TCP. This change requires change in create_ReAuthRequest() function in s1ap_utils.py and in the TCs test_attach_detach_rar_tcp_data.py and test_attach_detach_multiple_rar_tcp_data.py  . Shall I take this up in a separate PR?